ghc 9 - Couldn't match MonadCatch with MonadThrow

MonadCatch hasn't been touched since 2019 so this probably has something to do with ghc 9 changes.
```hs
[ 3 of 31] Compiling Polysemy.ConstraintAbsorber.MonadCatch
                            
/home/theo/prj/polysemy-zoo/src/Polysemy/ConstraintAbsorber/MonadCatch.hs:83:20: error:
    * Couldn't match type: C.MonadCatch (Sem r)
                     with: C.MonadThrow (Sem r)
      Expected: (C.MonadThrow (Sem r) => Sem r a) -> Sem r a
        Actual: (C.MonadCatch (Sem r) => Sem r a) -> Sem r a
    * In the expression: absorbMonadCatch
      In an equation for `absorbMonadThrow':
          absorbMonadThrow = absorbMonadCatch
    * Relevant bindings include
        absorbMonadThrow :: (C.MonadThrow (Sem r) => Sem r a) -> Sem r a
          (bound at src/Polysemy/ConstraintAbsorber/MonadCatch.hs:83:1)
   |                        
83 | absorbMonadThrow = absorbMonadCatch
   |                    ^^^^^^^^^^^^^^^^
```
The `stack.yaml` I used:
```yml
compiler: ghc-9.0.1
resolver: nightly-2020-05-07

allow-newer: true

packages:
- .

extra-deps:
- atomic-primops-0.8.4
- compact-0.2.0.0
- random-1.2.0
- splitmix-0.1.0.3
- syb-0.7.2.1
- th-abstraction-0.4.2.0
- QuickCheck-2.14.2
# GHC 9 patch (not yet on hackage)
- github: polysemy-research/polysemy
  commit: 09c65bdae9c9ce50a55c7ee2ab9681916f09f4c3
  subdirs: [.]
```

--- a/package.yaml
+++ b/package.yaml
@@ -31,7 +31,7 @@ dependencies:
 - reflection >= 2.1.4 && < 3.0.0
 - transformers >= 0.5.2.0 && < 0.6
 - text >= 1.1.0 && < 1.3
-- ghc-prim >= 0.5.2 && < 0.7
+- ghc-prim >= 0.5.2 && < 0.8
 - async >= 2.2 && < 3
 - streaming >= 0.2 && < 0.3
 - compact >= 0.1.0.1

--- a/polysemy-zoo.cabal
+++ b/polysemy-zoo.cabal
@@ -1,10 +1,10 @@
 cabal-version: 1.12
 
--- This file has been generated from package.yaml by hpack version 0.33.0.
+-- This file has been generated from package.yaml by hpack version 0.34.4.
 --
 -- see: https://github.com/sol/hpack
 --
--- hash: a5563140d5d76f52bdcb2c1def770fca8613155455f2e76e5286d6a5cd6505a3
+-- hash: 059b2cd3ed165884cb431bf574335de75fffc16b7f2c20755d5cf8881e475352
 
 name:           polysemy-zoo
 version:        0.7.0.1
@@ -63,7 +63,20 @@ library
       Paths_polysemy_zoo
   hs-source-dirs:
       src
-  default-extensions: DataKinds DeriveFunctor FlexibleContexts GADTs LambdaCase PolyKinds RankNTypes ScopedTypeVariables StandaloneDeriving TypeApplications TypeOperators TypeFamilies UnicodeSyntax
+  default-extensions:
+      DataKinds
+      DeriveFunctor
+      FlexibleContexts
+      GADTs
+      LambdaCase
+      PolyKinds
+      RankNTypes
+      ScopedTypeVariables
+      StandaloneDeriving
+      TypeApplications
+      TypeOperators
+      TypeFamilies
+      UnicodeSyntax
   ghc-options: -Wall
   build-depends:
       async >=2.2 && <3
@@ -73,12 +86,12 @@ library
     , containers >=0.5 && <0.7
     , contravariant <2
     , exceptions >=0.10.0 && <0.11
-    , ghc-prim >=0.5.2 && <0.7
+    , ghc-prim >=0.5.2 && <0.8
     , mtl >=2.0.1.0 && <3.0.0.0
     , polysemy >=1.2.1.0
     , random >=1.1 && <1.3
     , reflection >=2.1.4 && <3.0.0
-    , streaming >=0.2 && <0.3
+    , streaming ==0.2.*
     , text >=1.1.0 && <1.3
     , transformers >=0.5.2.0 && <0.6
   default-language: Haskell2010
@@ -99,7 +112,20 @@ test-suite polysemy-zoo-test
       Paths_polysemy_zoo
   hs-source-dirs:
       test
-  default-extensions: DataKinds DeriveFunctor FlexibleContexts GADTs LambdaCase PolyKinds RankNTypes ScopedTypeVariables StandaloneDeriving TypeApplications TypeOperators TypeFamilies UnicodeSyntax
+  default-extensions:
+      DataKinds
+      DeriveFunctor
+      FlexibleContexts
+      GADTs
+      LambdaCase
+      PolyKinds
+      RankNTypes
+      ScopedTypeVariables
+      StandaloneDeriving
+      TypeApplications
+      TypeOperators
+      TypeFamilies
+      UnicodeSyntax
   ghc-options: -Wall -threaded -rtsopts -with-rtsopts=-N -fplugin=Polysemy.Plugin
   build-tool-depends:
       hspec-discover:hspec-discover >=2.0
@@ -111,7 +137,7 @@ test-suite polysemy-zoo-test
     , containers >=0.5 && <0.7
     , contravariant <2
     , exceptions >=0.10.0 && <0.11
-    , ghc-prim >=0.5.2 && <0.7
+    , ghc-prim >=0.5.2 && <0.8
     , hspec >=2.6.0 && <3
     , mtl >=2.0.1.0 && <3.0.0.0
     , polysemy >=1.2.0.0
@@ -119,7 +145,7 @@ test-suite polysemy-zoo-test
     , polysemy-zoo
     , random >=1.1 && <1.3
     , reflection >=2.1.4 && <3.0.0
-    , streaming >=0.2 && <0.3
+    , streaming ==0.2.*
     , text >=1.1.0 && <1.3
     , transformers >=0.5.2.0 && <0.6
   default-language: Haskell2010

--- a/src/Polysemy/ConstraintAbsorber/MonadCatch.hs
+++ b/src/Polysemy/ConstraintAbsorber/MonadCatch.hs
@@ -80,7 +80,8 @@ absorbMonadThrow
        -- or 'C.MonadThrow' for
        -- 'Sem'. This might be something with type @'C.MonadCatch' e m => m a@.
   -> Sem r a
-absorbMonadThrow = absorbMonadCatch
+absorbMonadThrow =
+  absorbWithSem @C.MonadThrow @Action (CatchDict E.throw E.catch) (Sub Dict)
 {-# INLINABLE absorbMonadThrow #-}
 
 ------------------------------------------------------------------------------

--- a/src/Polysemy/ConstraintAbsorber/MonadCatch.hs
+++ b/src/Polysemy/ConstraintAbsorber/MonadCatch.hs
@@ -80,8 +80,7 @@ absorbMonadThrow
        -- or 'C.MonadThrow' for
        -- 'Sem'. This might be something with type @'C.MonadCatch' e m => m a@.
   -> Sem r a
-absorbMonadThrow =
-  absorbWithSem @C.MonadThrow @Action (CatchDict E.throw E.catch) (Sub Dict)
+absorbMonadThrow main = absorbMonadCatch main
 {-# INLINABLE absorbMonadThrow #-}
 
 ------------------------------------------------------------------------------

--- a/src/Polysemy/ConstraintAbsorber/MonadCont.hs
+++ b/src/Polysemy/ConstraintAbsorber/MonadCont.hs
@@ -28,7 +28,7 @@ absorbCont
        -- 'Sem'. This might be something with type @'C.MonadCont' m => m a@.
     -> Sem r a
 absorbCont = absorbWithSem @C.MonadCont @Action
-  (ContDict (callCC @ref))
+  (ContDict (\main -> callCC @ref $ \exit -> main exit))
   (Sub Dict)
 {-# INLINEABLE absorbCont #-}
 

--- a/src/Polysemy/IdempotentLowering.hs
+++ b/src/Polysemy/IdempotentLowering.hs
@@ -89,7 +89,7 @@ nat' f = pure $ coerce $ Nat' f
     -> base (∀ z. Sem (e ': r) z -> m z)
 fi .@! gi = do
   f <- fi
-  g <- gi f
+  g <- gi (\x -> f x)
   nat $ \z -> f $ g z
 infixl 8 .@!
 
@@ -112,7 +112,7 @@ infixl 8 .@!
     -> base (∀ z. Sem (e ': r) z -> m (f z))
 fi .@@! gi = do
   f <- fi
-  g <- gi f
+  g <- gi (\x -> f x)
   nat' $ \z -> f $ g z
 infixl 8 .@@!
 

